src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -72,195 +72,248 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
-                    catch (Exception ex)
-                    {
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{input}' to '{type.Name}'", ex);
-                    }
-
-                    if (conversionFailed)
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{input}' to '{type.Name}'");
+                    throw new InvalidVariableError(variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'", ex);
                 }
 
-                return;
+                if (ret == null)
+                    throw new InvalidVariableError(variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'");
+
+                return ret;
             }
 
-            if (type is ListGraphType listType)
+            static object ParseValueList(ListGraphType listGraphType, string variableName, object value)
             {
-                var listItemType = listType.ResolvedType;
+                if (value == null)
+                    return null;
 
-                if (input is IEnumerable list && !(input is string))
+                // note: a list can have a single child element which will automatically be intrepreted as a list of 1 elements. (see below rule)
+                // so, to prevent a string as being intrepreted as a list of chars (which get converted to strings), we ignore considering a string as an IEnumerable
+                if (value is IEnumerable values && !(value is string))
                 {
-                    var index = -1;
-                    foreach (var item in list)
-                        AssertValidVariableValue(schema, listItemType, item, $"{variableName}[{++index}]", hasDefaultValue);
+                    // create a list containing the parsed elements in the input list
+                    var valueOutputs = new List<object>(values is ICollection collection ? collection.Count : 0);
+                    int index = 0;
+                    foreach (var val in values)
+                    {
+                        // parse/validate values as required by graph type
+                        valueOutputs.Add(ParseValue(listGraphType.ResolvedType, variableName + "[" + index++ + "]", val));
+                    }
+                    return valueOutputs;
                 }
                 else
                 {
-                    AssertValidVariableValue(schema, listItemType, input, variableName, hasDefaultValue);
+                    // RULE: If the value passed as an input to a list type is not a list and not the null value,
+                    // then the result of input coercion is a list of size one, where the single item value is the
+                    // result of input coercion for the list’s item type on the provided value (note this may apply
+                    // recursively for nested lists).
+                    var result = ParseValue(listGraphType.ResolvedType, variableName, value);
```

Comment:
Does this variable name need index 0?

Reply:
I would not. It is used for the error message and would be a more accurate error message without it.

Reply:
OK